prow/spyglass/spyglass.go

```diff
@@ -0,0 +1,189 @@
+	Name string
+	// Title of view
+	Title string
+	// Html rendering of the view
+	HTMLView string
+	// Priority is the relative position of the view on the page
+	Priority int
+}
+
+// ViewRequest holds data sent by a view
+type ViewRequest struct {
+	ViewName    string              `json:"name"`
+	ViewerCache map[string][]string `json:"viewerCache"`
+	ViewData    string              `json:"viewData"`
+}
+
+// NewSpyglass constructs a spyglass object from a JobAgent
+func NewSpyglass(ja *jobs.JobAgent) *Spyglass {
+	return &Spyglass{
+		Lenses: make(map[string]Lens),
+		Ja:     ja,
+	}
+}
+
+// Views gets all views of all artifact files matching each regexp with a registered viewer
+func (s *Spyglass) Views(eyepiece map[string][]string) []Lens {
+	lenses := []Lens{}
+	for viewer, matches := range eyepiece {
+		if len(matches) != 0 {
+			title, err := viewers.Title(viewer)
+			if err != nil {
+				logrus.Error("Could not find artifact viewer with name ", viewer)
+			}
+			priority, err := viewers.Priority(viewer)
+			if err != nil {
+				logrus.Error("Could not find artifact viewer with name ", viewer)
+			}
+			lens := Lens{
+				Name:     viewer,
+				Title:    title,
+				HTMLView: "",
+				Priority: priority,
+			}
+			lenses = append(lenses, lens)
+			s.Lenses[viewer] = lens
+		}
+	}
+	// Make sure lenses are rendered in order by ascending priority
+	sort.Slice(lenses, func(i, j int) bool {
+		iname := lenses[i].Name
+		jname := lenses[j].Name
+		pi := lenses[i].Priority
+		pj := lenses[j].Priority
+		if pi == pj {
+			return iname < jname
+		}
+		return pi < pj
+	})
+	return lenses
+}
+
+// Refresh reloads the html view for a given set of objects
+func (s *Spyglass) Refresh(src string, podName string, viewReq *ViewRequest) Lens {
+	lens, ok := s.Lenses[viewReq.ViewName]
+	if !ok {
+		logrus.Errorf("Could not find Lens with name %s.", viewReq.ViewName)
+		return Lens{}
+	}
+	artifacts, err := s.FetchArtifacts(src, podName, viewReq.ViewerCache[viewReq.ViewName])
+	if err != nil {
+		logrus.WithError(err).Error("Error while fetching artifacts.")
+	}
+
+	view, err := viewers.View(viewReq.ViewName, artifacts, viewReq.ViewData)
+	if err != nil {
+		logrus.WithError(err).Error("Could not find a valid artifact viewer.")
+		return Lens{}
+	}
+	lens.HTMLView = view
+	return lens
+}
+
+// ListArtifacts handles muxing artifact sources to the correct fetcher implementations to list
+// all available artifact names
+func (s *Spyglass) ListArtifacts(src string, podName string) ([]string, error) {
+	artifacts := []string{}
+	var jobName string
+	var buildID string
+	// First check src
+	if isGCSSource(src) {
+		artifactFetcher := NewGCSArtifactFetcher()
+		gcsJobSource := NewGCSJobSource(src)
+		jobName = gcsJobSource.JobName()
+		buildID = gcsJobSource.BuildID()
+
+		artifacts = append(artifacts, artifactFetcher.Artifacts(gcsJobSource)...)
+
+	}
+	// Then check prowjob id for pod logs, pod spec, etc
+	if podName != "" || isProwJobSource(src) {
```

Comment:
if you provide podName and a GcsSource, the podlog artifacts will be process twice? no?

(same for `FetchArtifacts ` below)